### ISSUE TITLE
Very high RAM usage and compile time for serving big-ish schemas

### ISSUE THREAD

**Issue by akshaymankar** [higherkindness/mu-haskell#277]
While trying to write a service defined in a [not so big protobuf file](https://github.com/hashicorp/waypoint-plugin-sdk/blob/140d0145b90e3410546ba2702e4586fd445c066c/proto/plugin.proto), I noticed that the time it took to compile was very high, in this GHC also consumed too much RAM. I did some tests by creating couple of protobuf services and noticed that the compile time and GHC's RAM usage grows very quickly. The experiments and results can be found here: https://github.com/akshaymankar/mu-tests.

The compile time and RAM usage is not so bad when compiling with `-O0`, so I wouldn't think this is due to some type families going rogue, as I would hope that all the type checking still happens when compiling with `-O0`.

**Comment by serras** [comment]
Type checking still goes through, indeed. My guess is that GHC is doing a lot of inlining of the small type class instances, which makes the amount of code explode there.

One thing you could try is moving to `lts-17` to use GHC 8.10 instead of 8.8, since they've supposedly improved the story here.

**Comment by akshaymankar** [comment]
The example I made for this already uses GHC 8.10.3. I wanted to try GHC HEAD too, but there were significant compilation problems with the dependencies (I got stuck at `foundation`) as `base` has changed representation of all the `Word` and `Int` types, so I gave up. Next thing I'd try (if I find time!) is to reduce the dependency footprint of the example while still having high compile time and RAM usage, so it could be easier to use GHC HEAD.
